### ISSUE TITLE
Type is Never when pattern matching on type that inherits from str

### ISSUE THREAD

**Issue by marcinplatek** [microsoft/pyright#3883]
## Environment data

-   Language Server version: v2022.8.40
-   OS and version: macOS Monterey 12.5.1
-   Python version (& distribution if applicable, e.g. Anaconda): 3.10.6

## Code Snippet

```python
class A(str):
    def check(self):
        print(f"{self} works!")


def with_never(a: A):
    match a:
        case "TEST1" as m: # m has type never according to pylance
            m.check()
        case _:
            pass

def working(a: A):
    match a:
        case str("TEST2") as m:
            m.check()
        case _:
            pass

with_never(A("TEST1"))

working(A("TEST2"))


```

## Repro Steps

Set up pylance as strict in VSCode

The type will be `Never`, but the code will still work properly.

## Expected behavior

In both cases, m should be correctly typed as `A`.

## Actual behavior

<img width="646" alt="Screenshot 2022-08-31 at 13 40 18" src="https://user-images.githubusercontent.com/23441199/187670319-035ccdfc-e209-4c73-8783-f65535a85555.png">

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release of pyright.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.270, which I just published. It will also be included in a future release of pylance.
